data/random-teams.ts

```diff
@@ -1731,18 +1730,37 @@ export class RandomTeams {
 					if (!species.types.includes(type)) continue;
 				}
 			}
-			pokemonPool.push(pokemon);
-			baseSpeciesCount[species.baseSpecies] = (baseSpeciesCount[species.baseSpecies] || 0) + 1;
+
+			if (species.baseSpecies in pokemonPool) {
+				pokemonPool[species.baseSpecies].count++;
+				pokemonPool[species.baseSpecies].speciesArray.push(species.id);
+			} else {
+				pokemonPool[species.baseSpecies] = {count: 1, speciesArray: [species.id]};
+			}
 		}
 		// Include base species 1x if 1-3 formes, 2x if 4-6 formes, 3x if 7+ formes
-		for (const baseSpecies of Object.keys(baseSpeciesCount)) {
-			for (let i = 0; i < Math.min(Math.ceil(baseSpeciesCount[baseSpecies] / 3), 3); i++) {
-				baseSpeciesPool.push(baseSpecies);
-				// Squawkabilly has 4 formes, but only 2 functionally different formes, so only include it 1x
-				if (baseSpecies === 'Squawkabilly') break;
-			}
+		for (const baseSpecies of Object.keys(pokemonPool)) {
+			// Squawkabilly has 4 formes, but only 2 functionally different formes, so only include it 1x
+			if (baseSpecies === 'Squawkabilly') {
+				pokemonPool[baseSpecies].count = 1;
+			} else {
+				pokemonPool[baseSpecies].count = Math.min(Math.ceil(pokemonPool[baseSpecies].count / 3), 3);
+			}
+
+			/**
+			 * Weighted random shuffle
+			 * Uses the fact that for two uniform variables x1 and x2, x1^(1/w1) is larger than x2^(1/w2)
+			 * with probability equal to w1/(w1+w2), which is what we want. See e.g. here https://arxiv.org/pdf/1012.0256.pdf,
+			 * original paper is behind a paywall.
+			 */
+			const sortObject = {
+				baseSpecies: baseSpecies,
+				score: Math.pow(this.prng.next(), 1 / pokemonPool[baseSpecies].count),
+			};
```

Comment:
... where you only use it to calculate the `score` here.
```suggestion
			const sortObject = {
				baseSpecies: baseSpecies,
				score: Math.pow(this.prng.next(), 1 / weight),
			};
```